projects/packages/sync/src/class-actions.php

```diff
@@ -175,7 +175,9 @@ public static function add_sender_shutdown() {
 		) ) {
 			self::initialize_sender();
 			add_action( 'shutdown', array( self::$sender, 'do_sync' ), 9998 );
-			add_action( 'shutdown', array( self::$sender, 'do_full_sync' ), 9999 );
+			if ( self::should_initialize_sender( true ) ) {
```

Comment:
It feels weird to be checking against `should_initialize_sender` again even if we are doing it now with the `full_sync` arg to true. 
I guess the issue is that having the initial if altered by the `jetpack_sync_sender_should_load` filter makes things a bit more complicated. TBH I don't really understand the purpose of this filter.
Could we extract the logic of `Settings::is_dedicated_sync_enabled()` outside the `should_initialize_sender` so that we don't need to repeat everything?

Reply:
Yeah it's a bit awkward I also considered alternate approaches.

As you mentioned, we need to respect the `jetpack_sync_sender_should_load` filter.

Even if we extract the logic of `Settings::is_dedicated_sync_enabled()` outside the `should_initialize_sender`, we'd still have to repeat everything tho. 

Could you elaborate on the suggested approach?

Reply:
Something along these lines, but I don't love it either 
```diff
diff --git a/projects/packages/sync/src/class-actions.php b/projects/packages/sync/src/class-actions.php
index 2769030b4e..8f4b6e20d7 100644
--- a/projects/packages/sync/src/class-actions.php
+++ b/projects/packages/sync/src/class-actions.php
@@ -169,13 +169,15 @@ class Actions {
                 *
                 * @param bool should we load sync sender code for this request
                 */
+               $should_initialize_sender = self::should_initialize_sender();
+               $is_dedicated_sync        = Settings::is_dedicated_sync_enabled();
                if ( apply_filters(
                        'jetpack_sync_sender_should_load',
-                       self::should_initialize_sender()
+                       $is_dedicated_sync || $should_initialize_sender
                ) ) {
                        self::initialize_sender();
                        add_action( 'shutdown', array( self::$sender, 'do_sync' ), 9998 );
-                       if ( self::should_initialize_sender( true ) ) {
+                       if ( $should_initialize_sender ) {
                                add_action( 'shutdown', array( self::$sender, 'do_full_sync' ), 9999 );
                        }
                }
@@ -218,7 +220,7 @@ class Actions {
         *
         * @return bool
         */
-       public static function should_initialize_sender( $full_sync = false ) {
+       public static function should_initialize_sender() {
 
                // Allow for explicit disable of Sync from request param jetpack_sync_read_only.
                if ( isset( $_REQUEST['jetpack_sync_read_only'] ) ) { // phpcs:ignore WordPress.Security.NonceVerification
@@ -234,7 +236,7 @@ class Actions {
                 * For now, if dedicated Sync is enabled we will always initialize send, even for GET and unauthenticated requests
                 * but not for Full Sync, since it will still happen on shutdown.
                 */
-               if ( false === $full_sync && Settings::is_dedicated_sync_enabled() ) {
+               if ( Settings::is_dedicated_sync_enabled() ) {
                        return true;
                }
 
```

Reply:
Yeah this is what I wanted to avoid. Because this approach forces us to remember to add the Dedicated_Sync check whenever we used to check if sending is allowed.
Also I think in the above example you meant to fully remove the `Settings::is_dedicated_sync_enabled()` check from `should_initialize_sender`, right?

Reply:
>Yeah this is what I wanted to avoid. Because this approach forces us to remember to add the Dedicated_Sync check whenever we used to check if sending is allowed.

I get your point. Still feels pretty convoluted. But if we cannot find a simpler way I guess it's ok :shrug: